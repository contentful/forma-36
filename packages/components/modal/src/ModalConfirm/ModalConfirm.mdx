---
title: 'ModalConfirm'
type: 'component'
status: 'stable'
slug: /components/modal-confirm/
github: 'https://github.com/contentful/forma-36/tree/master/packages/components/modal/src'
storybook: 'https://f36-storybook.contentful.com/?path=/story/components-modal-modalconfirm--basic'
typescript: ./ModalConfirm.tsx
---

import { Props } from '@contentful/f36-docs-utils';

`ModalConfirm` is a convenience wrapper on top of `Modal`. It is intended to use for showing confirmation dialogs.

### Import

```js static=true
import { ModalConfirm } from '@contentful/f36-components';
// or
import { ModalConfirm } from '@contentful/f36-modal';
```

## Code examples

```jsx
function ModalConfirmExample() {
  const [isShown, setShown] = React.useState(false);
  return (
    <div>
      <Button variant="negative" onClick={() => setShown(true)}>
        Delete something
      </Button>
      <ModalConfirm
        isShown={isShown}
        onCancel={() => {
          setShown(false);
        }}
        onConfirm={() => {
          setShown(false);
        }}
      >
        <Text>Do you want to confirm your action?</Text>
      </ModalConfirm>
    </div>
  );
}
```

## Props

<Props of="ModalConfirm" />
