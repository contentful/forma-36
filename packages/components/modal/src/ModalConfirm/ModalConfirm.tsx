import React from 'react';

import { Modal } from '../CompoundModal';
import type { ModalProps } from '../Modal';
import type { ModalHeaderProps } from '../ModalHeader/ModalHeader';
import type { ModalContentProps } from '../ModalContent/ModalContent';
import type { ModalControlsProps } from '../ModalControls/ModalControls';
import { Button } from '@contentful/f36-button';

export interface ModalConfirmProps {
  /**
   * When true, the dialog is shown.
   */
  isShown: boolean;
  /**
   * Function that will be called when the confirm button is clicked. This does not close the ModalConfirm.
   */
  onConfirm(): void;
  /**
   * Function that will be called when the cancel button is clicked. This does not close the ModalConfirm.
   */
  onCancel: ModalProps['onClose'];
  /**
      Modal title that is used in header
    */
  title?: string;
  /**
   * Label of the confirm button
   */
  confirmLabel?: string | false;
  /**
   * Label of the cancel button
   */
  cancelLabel?: string | false;
  /**
   * The intent of the ModalConfirm. Used for the Button.
   */
  intent?: 'primary' | 'positive' | 'negative';
  /**
   * Size of the modal window
   */
  size?: ModalProps['size'];
  /**
   * Boolean indicating if clicking the overlay should close the overlay.
   */
  shouldCloseOnOverlayClick?: boolean;
  /**
   * Boolean indicating if pressing the esc key should close the overlay.
   */
  shouldCloseOnEscapePress?: boolean;
  /**
   * When true, the confirm button is set to disabled.
   */
  isConfirmDisabled?: boolean;
  /**
   * When true, the confirm button is set to loading.
   */
  isConfirmLoading?: boolean;
  /**
   * Are modals higher than viewport allowed
   */
  allowHeightOverflow?: boolean;

  /**
   * Optional props to override ModalHeader behaviour
   */
  modalHeaderProps?: Partial<ModalHeaderProps>;

  /**
   * Optional props to override ModalContent behaviour
   */
  modalContentProps?: Partial<ModalContentProps>;

  /**
   * Optional props to override ModalControl behaviour
   */
  modalControlsProps?: Partial<ModalControlsProps>;

  /**
   * Optional property to set initial focus
   */
  initialFocusRef?: React.RefObject<HTMLElement>;

  testId?: string;
  confirmTestId?: string;
  cancelTestId?: string;
  children: React.ReactNode;
}

export const ModalConfirm = ({
  allowHeightOverflow = false,
  cancelLabel = 'Cancel',
  cancelTestId = 'cf-ui-modal-confirm-cancel-button',
  children,
  confirmLabel = 'Confirm',
  confirmTestId = 'cf-ui-modal-confirm-confirm-button',
  intent = 'positive',
  isConfirmDisabled = false,
  isConfirmLoading = false,
  isShown,
  modalContentProps,
  modalControlsProps,
  modalHeaderProps,
  onCancel,
  onConfirm,
  shouldCloseOnEscapePress = true,
  shouldCloseOnOverlayClick = true,
  size = 'medium',
  testId = 'cf-ui-modal-confirm',
  title = 'Are you sure?',
  initialFocusRef,
}: ModalConfirmProps) => {
  const cancelRef = React.useRef(null);

  const confirmButton = confirmLabel ? (
    <Button
      testId={confirmTestId}
      isDisabled={isConfirmDisabled}
      isLoading={isConfirmLoading}
      variant={intent}
      size="small"
      onClick={() => onConfirm()}
    >
      {confirmLabel}
    </Button>
  ) : null;

  const cancelButton = cancelLabel ? (
    <Button
      testId={cancelTestId}
      variant="secondary"
      onClick={onCancel}
      size="small"
      ref={initialFocusRef || cancelRef}
    >
      {cancelLabel}
    </Button>
  ) : null;

  return (
    <Modal
      testId={testId}
      isShown={isShown}
      onClose={onCancel}
      size={size}
      shouldCloseOnOverlayClick={shouldCloseOnOverlayClick}
      shouldCloseOnEscapePress={shouldCloseOnEscapePress}
      allowHeightOverflow={allowHeightOverflow}
      initialFocusRef={cancelRef}
    >
      {() => {
        return (
          <React.Fragment>
            <Modal.Header title={title || ''} {...modalHeaderProps} />
            <Modal.Content {...modalContentProps}>{children}</Modal.Content>
            <Modal.Controls {...modalControlsProps}>
              {cancelButton}
              {confirmButton}
            </Modal.Controls>
          </React.Fragment>
        );
      }}
    </Modal>
  );
};

ModalConfirm.displayName = 'ModalConfirm';
