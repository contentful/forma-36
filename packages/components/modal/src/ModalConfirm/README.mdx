---
title: 'ModalConfirm'
type: 'component'
status: 'stable'
section: 'modalComponents'
slug: /components/modal-confirm/
github: 'https://github.com/contentful/forma-36/tree/master/packages/components/modal/src'
typescript: ./ModalConfirm.tsx
---

`ModalConfirm` is a convenience wrapper on top of [Modal]. It is intended to use for showing confirmation dialogs.

## Import

```js static=true
import { ModalConfirm } from '@contentful/f36-components';
// or
import { ModalConfirm } from '@contentful/f36-modal';
```

## Examples

### Basic

`ModalConfirm` has 3 different intents: `positive`, `primary` and `negative`. Changing this property changes the variant of the primary button.

Here is an example of a modal with a positive intent:

```jsx file=../../examples/ModalConfirmBasicExample.tsx

```

### Changing button labels

You can change action button labels by using `cancelLabel` and `confirmLabel`.

```jsx file=../../examples/ModalConfirmPromptExample.tsx

```

### Changing sizes

There are 4 different sizes for modals: `small`, `medium`, `large`, `fullWidth`. Pick a size based on the amount of content you have to present in them.

`ModalConfirm` behavior is exacly the same as [Modal]. Please, refer to [Modal] documentation for code example.

### Using ModalLauncher

When you want to open a modal from a function (outside of React component code), use [ModalLauncher.](../components/modal-launcher)

## Props (API reference)

<PropsTable of="ModalConfirm" />

## Content guidelines

Please, refer to [Modal] documentation.

## Accessibility

- When the modal opens, focus is trapped within it.
- When the modal opens, focus is automatically set to the `Cancel` button, or the element from `initialFocusRef`.
- When the modal closes, focus returns to the element that was focused before the modal activated.
- By default, clicking on the overlay closes the modal.
- By default, pressing `Esc` closes the modal.
- Scrolling is blocked on the elements behind the modal.

[modal]: ../components/modal
