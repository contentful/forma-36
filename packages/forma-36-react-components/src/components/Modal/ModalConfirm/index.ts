export * from './ModalConfirm';
export { default } from './ModalConfirm';
