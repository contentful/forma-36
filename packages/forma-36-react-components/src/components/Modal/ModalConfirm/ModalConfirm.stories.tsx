import React, { useState } from 'react';
import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import { text, boolean, select } from '@storybook/addon-knobs';

import ModalConfirm from './ModalConfirm';
import Button from '../../Button';
import TextInput from '../../TextInput';

function DefaultStory() {
  const [isShown, setShown] = useState(false);
  return (
    <div>
      <Button buttonType="negative" onClick={() => setShown(true)}>
        Delete something
      </Button>
      <ModalConfirm
        isShown={isShown}
        title={text('title', ModalConfirm.defaultProps.title)}
        intent={select(
          'intent',
          ['negative', 'positive', 'primary'],
          ModalConfirm.defaultProps.intent,
        )}
        secondaryIntent={select('secondaryIntent', [
          'negative',
          'positive',
          'primary',
          'muted',
        ])}
        size={select(
          'size',
          ['small', 'medium', 'large', '300px', '1500px'],
          ModalConfirm.defaultProps.size,
        )}
        shouldCloseOnEscapePress={boolean(
          'shouldCloseOnEscapePress',
          ModalConfirm.defaultProps.shouldCloseOnEscapePress,
        )}
        shouldCloseOnOverlayClick={boolean(
          'shouldCloseOnOverlayClick',
          ModalConfirm.defaultProps.shouldCloseOnOverlayClick,
        )}
        confirmLabel={text(
          'confirmLabel',
          ModalConfirm.defaultProps.confirmLabel,
        )}
        secondaryLabel={text('secondaryLabel')}
        cancelLabel={text('cancelLabel', ModalConfirm.defaultProps.cancelLabel)}
        isConfirmDisabled={boolean(
          'isConfirmDisabled',
          ModalConfirm.defaultProps.isConfirmDisabled,
        )}
        isSecondaryDisabled={boolean('secondaryDisabled')}
        isConfirmLoading={boolean(
          'isConfirmLoading',
          ModalConfirm.defaultProps.isConfirmLoading,
        )}
        isSecondaryLoading={boolean('isSecondaryLoading')}
        testId={text('testId', ModalConfirm.defaultProps.testId)}
        confirmTestId={text(
          'confirmTestId',
          ModalConfirm.defaultProps.confirmTestId,
        )}
        secondaryTestId={text(
          'secondaryTestId',
          ModalConfirm.defaultProps.secondaryTestId,
        )}
        cancelTestId={text(
          'cancelTextId',
          ModalConfirm.defaultProps.cancelTestId,
        )}
        onCancel={() => {
          setShown(false);
          action('onCancel')();
        }}
        onConfirm={() => {
          setShown(false);
          action('onConfirm')();
        }}
        onSecondary={() => {
          setShown(false);
          action('onSecondary')();
        }}
      >
        <p>You are about to delete SOMETHING. Think twice!</p>
      </ModalConfirm>
    </div>
  );
}

function ComplexStory() {
  const [isShown, setShown] = useState(false);
  const [isLoading, setLoading] = useState(false);
  const [repeat, setRepeat] = useState('');

  return (
    <div>
      <Button buttonType="negative" onClick={() => setShown(true)}>
        Delete something
      </Button>
      <ModalConfirm
        isShown={isShown}
        intent="negative"
        isConfirmDisabled={repeat !== 'unlock'}
        isConfirmLoading={isLoading}
        onCancel={() => {
          setShown(false);
          action('onCancel')();
        }}
        onConfirm={() => {
          setLoading(true);
          setTimeout(() => {
            setLoading(false);
            setShown(false);
            setRepeat('');
          }, 1500);
          action('onConfirm')();
        }}
      >
        <p>
          Type <strong>unlock</strong> to allow confirming this modal
        </p>
        <TextInput
          value={repeat}
          onChange={(e) => setRepeat((e.target as HTMLInputElement).value)}
        />
      </ModalConfirm>
    </div>
  );
}

storiesOf('Components/Modal/ModalConfirm', module)
  .addParameters({
    propTypes: ModalConfirm['__docgenInfo'],
    component: ModalConfirm,
  })
  .add('default', () => <DefaultStory />)
  .add('complex example', () => <ComplexStory />);
