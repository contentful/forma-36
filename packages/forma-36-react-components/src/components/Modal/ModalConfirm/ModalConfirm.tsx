import React, { Component } from 'react';

import Modal, { ModalSizeType } from '../Modal/Modal';
import { ModalHeaderProps } from '../ModalHeader/ModalHeader';
import { ModalContentProps } from '../ModalContent/ModalContent';
import { ModalControlsProps } from '../ModalControls/ModalControls';
import Button from '../../Button';

export type ModalConfirmProps = {
  /**
   * When true, the dialog is shown.
   */
  isShown: boolean;
  /**
   * Function that will be called when the confirm button is clicked. This does not close the ModalConfirm.
   */
  onConfirm: Function;
  /**
   * Function that will be called when the cancel button is clicked. This does not close the ModalConfirm.
   */
  onCancel: Function;
  /**
      Modal title that is used in header
    */
  title?: string;
  /**
   * Label of the confirm button
   */
  confirmLabel?: string;
  /**
   * Label of the cancel button
   */
  cancelLabel?: string;
  /**
   * The intent of the ModalConfirm. Used for the Button.
   */
  intent?: 'primary' | 'positive' | 'negative';
  /**
      Size of the modal window
    */
  size?: ModalSizeType;
  /**
   * Boolean indicating if clicking the overlay should close the overlay.
   */
  shouldCloseOnOverlayClick?: boolean;
  /**
   * Boolean indicating if pressing the esc key should close the overlay.
   */
  shouldCloseOnEscapePress?: boolean;
  /**
   * When true, the confirm button is set to disabled.
   */
  isConfirmDisabled?: boolean;
  /**
   * When true, the confirm button is set to loading.
   */
  isConfirmLoading?: boolean;

  /**
   * Optional props to override ModalHeader behaviour
   */
  modalHeaderProps?: Partial<ModalHeaderProps>;

  /**
   * Optional props to override ModalContent behaviour
   */
  modalContentProps?: Partial<ModalContentProps>;

  /**
   * Optional props to override ModalControl behaviour
   */
  modalControlsProps?: Partial<ModalControlsProps>;

  testId?: string;
  confirmTestId?: string;
  cancelTestId?: string;
  children: React.ReactNode;
} & typeof defaultProps;

const defaultProps = {
  testId: 'cf-ui-modal-confirm',
  confirmTestId: 'cf-ui-modal-confirm-confirm-button',
  cancelTestId: 'cf-ui-modal-confirm-cancel-button',
  title: 'Are you sure?',
  confirmLabel: 'Confirm',
  cancelLabel: 'Cancel',
  intent: 'positive',
  shouldCloseOnOverlayClick: true,
  shouldCloseOnEscapePress: true,
  isConfirmDisabled: false,
  isConfirmLoading: false,
  size: 'medium',
};

export class ModalConfirm extends Component<ModalConfirmProps> {
  static defaultProps = defaultProps;

  render() {
    const {
      children,
      testId,
      isShown,
      title,
      onConfirm,
      onCancel,
      size,
      confirmLabel,
      cancelLabel,
      intent,
      shouldCloseOnOverlayClick,
      shouldCloseOnEscapePress,
      isConfirmDisabled,
      isConfirmLoading,
      confirmTestId,
      cancelTestId,
    } = this.props;

    return (
      <Modal
        testId={testId}
        isShown={isShown}
        onClose={onCancel}
        size={size}
        shouldCloseOnOverlayClick={shouldCloseOnOverlayClick}
        shouldCloseOnEscapePress={shouldCloseOnEscapePress}
      >
        {() => (
          <div>
            <Modal.Header title={title} {...this.props.modalHeaderProps} />
            <Modal.Content {...this.props.modalContentProps}>
              {children}
            </Modal.Content>
            <Modal.Controls {...this.props.modalControlsProps}>
              <Button
                testId={confirmTestId}
                disabled={isConfirmDisabled}
                loading={isConfirmLoading}
                buttonType={intent}
                onClick={() => onConfirm()}
              >
                {confirmLabel}
              </Button>
              <Button
                testId={cancelTestId}
                buttonType="muted"
                onClick={() => onCancel()}
              >
                {cancelLabel}
              </Button>
            </Modal.Controls>
          </div>
        )}
      </Modal>
    );
  }
}

export default ModalConfirm;
