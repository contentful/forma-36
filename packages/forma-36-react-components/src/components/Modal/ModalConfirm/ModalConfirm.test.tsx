import React from 'react';
import { mount } from 'enzyme';
import axe from '../../../utils/axeHelper';
import ModalConfirm from './ModalConfirm';

jest.mock(
  'react-modal',
  () =>
    // eslint-disable-next-line
    function ReactModalMock({ children }: { children: React.ReactNode }) {
      return <div className="react-modal">{children}</div>;
    },
);

jest.mock(
  '../../Button',
  () =>
    function Button() {
      return null;
    },
);

it('renders the component', () => {
  const output = mount(
    <ModalConfirm isShown onConfirm={() => {}} onCancel={() => {}}>
      ModalConfirm
    </ModalConfirm>,
  );

  expect(output.find('[data-test-id="cf-ui-modal-confirm"]')).toMatchSnapshot();
});

it('renders the component with override header, content and controls', () => {
  const output = mount(
    <ModalConfirm
      isShown
      onConfirm={() => {}}
      onCancel={() => {}}
      modalHeaderProps={{
        isNotWrapped: true,
        className: 'additional header class',
      }}
      modalContentProps={{
        className: 'additional content class',
      }}
      modalControlsProps={{
        className: 'additional controls class',
      }}
    >
      ModalConfirm
    </ModalConfirm>,
  );

  expect(output.find('[data-test-id="cf-ui-modal-confirm"]')).toMatchSnapshot();
});

it('can accept custom labels', () => {
  const output = mount(
    <ModalConfirm
      isShown
      intent="negative"
      confirmLabel="Yes, delete"
      secondaryLabel="Disable, instead"
      cancelLabel="No, I changed my mind"
      onConfirm={() => {}}
      onSecondary={() => {}}
      onCancel={() => {}}
    >
      ModalConfirm
    </ModalConfirm>,
  );

  expect(output.find('[data-test-id="cf-ui-modal-confirm"]')).toMatchSnapshot();
});

it('has no a11y issues', async () => {
  const output = mount(
    <ModalConfirm isShown onConfirm={() => {}} onCancel={() => {}}>
      ModalConfirm
    </ModalConfirm>,
  ).html();
  const results = await axe(output);

  expect(results).toHaveNoViolations();
});
