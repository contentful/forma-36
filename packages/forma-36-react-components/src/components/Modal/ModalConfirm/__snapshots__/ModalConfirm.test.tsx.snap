// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can accept custom labels 1`] = `
<div
  className="Modal"
  data-test-id="cf-ui-modal-confirm"
  style={
    Object {
      "width": "520px",
    }
  }
>
  <div>
    <ModalHeader
      testId="cf-ui-modal-header"
      title="Are you sure?"
    >
      <div
        className="ModalHeader"
        data-test-id="cf-ui-modal-header"
      >
        <h1
          className="ModalHeader__title"
        >
          Are you sure?
        </h1>
      </div>
    </ModalHeader>
    <ModalContent
      testId="cf-ui-modal-content"
    >
      <div
        className="ModalContent"
        data-test-id="cf-ui-modal-content"
      >
        ModalConfirm
      </div>
    </ModalContent>
    <ModalControls
      testId="cf-ui-modal-controls"
    >
      <div
        className="ModalControls"
        data-test-id="cf-ui-modal-controls"
      >
        <Button
          buttonType="negative"
          disabled={false}
          loading={false}
          onClick={[Function]}
          testId="cf-ui-modal-confirm-confirm-button"
        />
        <Button
          buttonType="muted"
          onClick={[Function]}
          testId="cf-ui-modal-confirm-cancel-button"
        />
      </div>
    </ModalControls>
  </div>
</div>
`;

exports[`renders the component 1`] = `
<div
  className="Modal"
  data-test-id="cf-ui-modal-confirm"
  style={
    Object {
      "width": "520px",
    }
  }
>
  <div>
    <ModalHeader
      testId="cf-ui-modal-header"
      title="Are you sure?"
    >
      <div
        className="ModalHeader"
        data-test-id="cf-ui-modal-header"
      >
        <h1
          className="ModalHeader__title"
        >
          Are you sure?
        </h1>
      </div>
    </ModalHeader>
    <ModalContent
      testId="cf-ui-modal-content"
    >
      <div
        className="ModalContent"
        data-test-id="cf-ui-modal-content"
      >
        ModalConfirm
      </div>
    </ModalContent>
    <ModalControls
      testId="cf-ui-modal-controls"
    >
      <div
        className="ModalControls"
        data-test-id="cf-ui-modal-controls"
      >
        <Button
          buttonType="positive"
          disabled={false}
          loading={false}
          onClick={[Function]}
          testId="cf-ui-modal-confirm-confirm-button"
        />
        <Button
          buttonType="muted"
          onClick={[Function]}
          testId="cf-ui-modal-confirm-cancel-button"
        />
      </div>
    </ModalControls>
  </div>
</div>
`;

exports[`renders the component with override header, content and controls 1`] = `
<div
  className="Modal"
  data-test-id="cf-ui-modal-confirm"
  style={
    Object {
      "width": "520px",
    }
  }
>
  <div>
    <ModalHeader
      className="additional header class"
      isNotWrapped={true}
      testId="cf-ui-modal-header"
      title="Are you sure?"
    >
      <div
        className="ModalHeader additional header class"
        data-test-id="cf-ui-modal-header"
      >
        <h1
          className="ModalHeader__title ModalHeader__title--is-not-wrapped"
        >
          Are you sure?
        </h1>
      </div>
    </ModalHeader>
    <ModalContent
      className="additional content class"
      testId="cf-ui-modal-content"
    >
      <div
        className="ModalContent additional content class"
        data-test-id="cf-ui-modal-content"
      >
        ModalConfirm
      </div>
    </ModalContent>
    <ModalControls
      className="additional controls class"
      testId="cf-ui-modal-controls"
    >
      <div
        className="ModalControls additional controls class"
        data-test-id="cf-ui-modal-controls"
      >
        <Button
          buttonType="positive"
          disabled={false}
          loading={false}
          onClick={[Function]}
          testId="cf-ui-modal-confirm-confirm-button"
        />
        <Button
          buttonType="muted"
          onClick={[Function]}
          testId="cf-ui-modal-confirm-cancel-button"
        />
      </div>
    </ModalControls>
  </div>
</div>
`;
